Multiple configurations with shared schemas directory: index.ts overwritten by last config

## Description

**⚠️ This is a v8 regression** - the same configuration works correctly in Orval 7.17.0 but breaks in 8.0.3.

When using multiple Orval configurations that write to the same `schemas` directory, the `index.ts` file is overwritten by the last configuration to run, causing only the last configuration's models to be exported. This results in build errors when earlier configurations try to import their models from the shared index.

## Current Behavior

With two configurations (`commandApi` and `queryApi`) writing to the same schemas directory:

1. `commandApi` generates:
   - Command model files in `src/api/models/`
   - `index.ts` exporting command models
   - `commands.ts` importing from `./models`

2. `queryApi` generates:
   - Query model files in `src/api/models/`
   - **OVERWRITES** `index.ts` with ONLY query model exports
   - `queries.ts` importing from `./models`

3. Result:
   - ✅ All model files exist (146 files in our case)
   - ❌ `index.ts` only exports query models
   - ❌ `commands.ts` imports fail with "Module has no exported member" errors

## Expected Behavior

One of the following:
1. `index.ts` should contain exports from ALL configurations that wrote to the directory (merged/accumulated exports)
2. Configuration option to control index generation strategy (e.g., `indexMergeStrategy: 'append' | 'overwrite'`)
3. Official documentation describing the recommended pattern for this use case
4. Clear error/warning when multiple configs target the same schemas directory

## Minimal Reproduction

### Project Structure
```
my-project/
├── openapi-commands.json    # Command API spec
├── openapi-queries.json     # Query API spec
└── orval.config.js
```

### orval.config.js
```javascript
module.exports = {
  commandApi: {
    output: {
      mode: 'split',
      target: 'src/api/commands.ts',
      schemas: 'src/api/models',        // ← SHARED DIRECTORY
      client: 'react-query',
      indexFiles: true,
    },
    input: {
      target: './openapi-commands.json',
    },
  },
  queryApi: {
    output: {
      mode: 'split',
      target: 'src/api/queries.ts',
      schemas: 'src/api/models',        // ← SAME SHARED DIRECTORY
      client: 'react-query',
      indexFiles: true,
    },
    input: {
      target: './openapi-queries.json',
    },
  },
};
```

### Example OpenAPI Specs

**openapi-commands.json** (simplified):
```json
{
  "openapi": "3.0.0",
  "paths": {
    "/api/commands/create-user": {
      "post": {
        "operationId": "createUser",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": { "$ref": "#/components/schemas/CreateUserRequest" }
            }
          }
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": { "$ref": "#/components/schemas/CreateUserResponse" }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "CreateUserRequest": {
        "type": "object",
        "properties": {
          "username": { "type": "string" },
          "email": { "type": "string" }
        }
      },
      "CreateUserResponse": {
        "type": "object",
        "properties": {
          "userId": { "type": "string" }
        }
      }
    }
  }
}
```

**openapi-queries.json** (simplified):
```json
{
  "openapi": "3.0.0",
  "paths": {
    "/api/queries/get-user/{id}": {
      "get": {
        "operationId": "getUser",
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": { "type": "string" }
          }
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": { "$ref": "#/components/schemas/UserDto" }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "UserDto": {
        "type": "object",
        "properties": {
          "id": { "type": "string" },
          "username": { "type": "string" },
          "email": { "type": "string" }
        }
      }
    }
  }
}
```

### Steps to Reproduce

1. Create the files above
2. Run `npx orval`
3. Check `src/api/models/index.ts`

**Result**: Only contains exports for `UserDto` (from queryApi), missing `CreateUserRequest` and `CreateUserResponse` (from commandApi)

4. Run `tsc` or your build tool

**Result**: Build errors like:
```
src/api/commands.ts:5:3 - error TS2305: Module '"./models"' has no exported member 'CreateUserRequest'.
```

## Regression Evidence

**Orval 7.17.0**: ✅ Works correctly - `index.ts` exports models from BOTH configs
**Orval 8.0.3**: ❌ Broken - `index.ts` only exports last config's models

This is a breaking change that blocks upgrades for CQRS architectures and other multi-config setups.

## Use Case / Context

This scenario is common in CQRS (Command Query Responsibility Segregation) architectures where:
- Commands and queries are separate OpenAPI specifications
- Both share common domain models
- Both should export to a single shared models directory for DRY principles

Workarounds like separate model directories (`command-models/` and `query-models/`) defeat the purpose of having shared domain models and require updating all import paths throughout the codebase.

## Related Issues/Discussions

- [Discussion #756](https://github.com/orval-labs/orval/discussions/756) - Similar issue reported
- [Issue #2044](https://github.com/orval-labs/orval/issues/2044) - Index file customization (may be related)

## Environment

- **Orval Version**: 8.0.3 (regression from 7.17.0)
- **Node Version**: 22.14.0
- **Package Manager**: npm 10.9.2

## Proposed Solutions

### Option 1: Accumulative Index Generation
Add a configuration option:
```javascript
output: {
  schemas: 'src/api/models',
  indexFiles: true,
  indexMergeStrategy: 'append', // or 'overwrite' (default, current behavior)
}
```

### Option 2: Post-Generation Hook Support
Document official pattern using `afterAllFilesWrite` hook to merge index files (currently undocumented community workaround)

### Option 3: Workspace-Level Index
Allow a root-level configuration that generates a merged index.ts for multiple sub-configs:
```javascript
module.exports = {
  workspace: 'src/api',
  sharedSchemas: 'models', // Creates merged index.ts
  configs: {
    commandApi: { /* ... */ },
    queryApi: { /* ... */ },
  }
};
```

## Additional Notes

- All individual model files ARE generated correctly (no duplicates, no overwrites)
- The ONLY issue is the `index.ts` file being overwritten instead of merged
- This worked correctly in Orval 7, making this a high-severity regression
- Community workaround exists using `afterAllFilesWrite` hook with custom merge script, but it's fragile and unsupported

Thank you for considering this issue!

--- a/packages/core/src/writers/schemas.test.ts
+++ b/packages/core/src/writers/schemas.test.ts
@@ -616,7 +616,7 @@ describe('fixRegularSchemaImports', () => {
 });
 
 describe('writeSchemas indexFiles', () => {
-  it('replaces index exports with the current run when re-generating the same schema path', async () => {
+  it('merges index exports across projects that share a schema path (#2842)', async () => {
     const tempDir = await fs.mkdtemp(
       path.join(os.tmpdir(), 'orval-schema-index-'),
     );
@@ -652,10 +652,10 @@ describe('writeSchemas indexFiles', () => {
       const indexPath = path.join(schemaPath, 'index.ts');
       const content = await fs.readFile(indexPath, 'utf8');
 
+      expect(content).toContain("export * from './createUserRequest';");
+      expect(content).toContain("export * from './createUserResponse';");
       expect(content).toContain("export * from './userDto';");
       expect(content).toContain("export * from './userListResponse';");
-      expect(content).not.toContain("export * from './createUserRequest';");
-      expect(content).not.toContain("export * from './createUserResponse';");
     } finally {
       await fs.remove(tempDir);
     }

--- a/packages/core/src/writers/schemas.test.ts
+++ b/packages/core/src/writers/schemas.test.ts
@@ -652,10 +652,15 @@ describe('writeSchemas indexFiles', () => {
       const indexPath = path.join(schemaPath, 'index.ts');
       const content = await fs.readFile(indexPath, 'utf8');
 
-      expect(content).toContain("export * from './createUserRequest';");
-      expect(content).toContain("export * from './createUserResponse';");
-      expect(content).toContain("export * from './userDto';");
-      expect(content).toContain("export * from './userListResponse';");
+      const exportLines = content
+        .split('\n')
+        .filter((line) => line.startsWith('export * from'));
+      expect(exportLines).toEqual([
+        "export * from './createUserRequest';",
+        "export * from './createUserResponse';",
+        "export * from './userDto';",
+        "export * from './userListResponse';",
+      ]);
     } finally {
       await fs.remove(tempDir);
     }

--- a/packages/core/src/writers/schemas.ts
+++ b/packages/core/src/writers/schemas.ts
@@ -615,7 +615,14 @@ export async function writeSchemas({
         }
       }
 
-      const exports = [...new Set(currentExports)]
+      const existingContent = await fs.readFile(schemaFilePath, 'utf8');
+      const existingExports = [
+        ...existingContent.matchAll(
+          /^\s*export\s+\*\s+from\s+['"]([^'"]+)['"]\s*;?\s*$/gm,
+        ),
+      ].map(([, specifier]) => `export * from '${specifier}';`);
+
+      const exports = [...new Set([...existingExports, ...currentExports])]
         .toSorted((a, b) => a.localeCompare(b, 'en', { numeric: true }))
         .join('\n');
 

